src/test/clojurescript/clara/test_salience.cljs

```diff
@@ -0,0 +1,115 @@
+  [Temperature]
+  =>
+  (swap! salience-rule-output conj 50))
+
+(defrule salience-rule3
+  {:salience 0}
+  [Temperature]
+  =>
+  (swap! salience-rule-output conj 0))
+
+(defrule salience-rule4
+  {:salience -50}
+  [Temperature]
+  =>
+  (swap! salience-rule-output conj -50))
+
+
+(deftest test-salience
+  (doseq [[sort-fn
+           group-fn
+           expected-order]
+
+          [[:default-sort :default-group :forward-order]
+           [:default-sort :salience-group :forward-order]
+           [:default-sort :neg-salience-group :backward-order]
+
+           [:numeric-greatest-sort :default-group :forward-order]
+           [:numeric-greatest-sort :salience-group :forward-order]
+           [:numeric-greatest-sort :neg-salience-group :backward-order]
+
+
+           [:boolean-greatest-sort :default-group :forward-order]
+           [:boolean-greatest-sort :salience-group :forward-order]
+           [:boolean-greatest-sort :neg-salience-group :backward-order]
+
+
+           [:numeric-least-sort :default-group :backward-order]
+           [:numeric-least-sort :salience-group :backward-order]
+           [:numeric-least-sort :neg-salience-group :forward-order]
+
+           [:boolean-least-sort :default-group :backward-order]
+           [:boolean-least-sort :salience-group :backward-order]
+           [:boolean-least-sort :neg-salience-group :forward-order]]]
+
+    (let [numeric-greatest-sort (fn [x y]
+                                  (cond
+                                    (= x y) 0
+                                    (> x y) -1
+                                    :else 1))
+
+          numeric-least-sort (fn [x y]
+                               (numeric-greatest-sort y x))
+
+          salience-group-fn (fn [production]
+                              (or (some-> production :props :salience)
+                                0))
+
+          neg-salience-group-fn (fn [p]
+                                  (- (salience-group-fn p)))]
+
+      (defsession test-salience-session 'clara.test-salience
+                     :cache false
+                     :activation-group-sort-fn (condp = sort-fn
+                                                 :default-sort nil
+                                                 :numeric-greatest-sort numeric-greatest-sort
+                                                 :numeric-least-sort numeric-least-sort
+                                                 :boolean-greatest-sort  >
+                                                 :boolean-least-sort <)
+                     :activation-group-fn (condp = group-fn
+                                            :default-group nil
+                                            :salience-group salience-group-fn
+                                            :neg-salience-group neg-salience-group-fn))
+
+      ;; Ensure the rule output reflects the salience-defined order.
+      ;; independently of the order of the rules.
+      (dotimes [n 10]
```

Comment:
I don't see any variation of the rule order happening here, as suggested by the comment above.  I don't think varying the order is really necessary though since you have tests where the ordering is expected to be "backward" so if Clara were just firing these rules in their order they are defined the test would fail.  A few quick experiments suggest that defining sessions based on individual rules doesn't work in ClojureScript, which makes some sense to me based on the compilation model in ClojureScript, although I'd need to dig further into it to be sure I'm not just missing something.  If this is true, you'd need to create different namespace sources to vary the rule order, but I don't think that's really necessary since you test the "backwards" case.

I also don't see what the dotimes is accomplishing - it looks to me like it is just executing the same test case 10 times.  Am I missing something?